### ISSUE TITLE
Boot counting broken?

### ISSUE THREAD

**Issue by poidl** [systemd/systemd#15667]
Describe the bug

Boot counting seems to be broken (Arch linux package systemd v245.5-2). I have loader files

```
arch-v1.conf
arch-v2+1.conf
```
After adding `arch-v2+1.conf`, `bootctl` lists it as the default entry. It reboots into that entry and if the boot fails the file is renamed correctly to `arch-v2+0-1.conf`. However, `bootctl` now lists 
`arch-v2+0-1.conf` as the default entry, and if I reboot and it fails again, it renames it to  `arch-v2+0-2.conf` and after another reboot to `arch-v2+0-3.conf` and so on. Instead it should switch back to `arch-v1.conf`, right?

To Reproduce

I use a `loader.conf` with just a `timeout` entry. Place `arch-v1.conf` and `arch-v2+1.conf` into the `/loader/entries/` directory. Then let `arch-v2+1.conf` fail by any mechanism. I use

```
tee /etc/systemd/system/deadmansswitch.service << EOF
[Install]
RequiredBy=boot-complete.target

[Unit]
Before=boot-complete.target

[Service]
Type=oneshot
ExecStartPre=/bin/sleep 60
ExecStart=/bin/ls /tmp/deadmansswitch
EOF

systemctl enable deadmansswitch
```

**Comment by poettering** [comment]
hmm, let me get this right, you have the same boot menu entry with and without boot counting? It's not how this is supposed to be used. Either you have bootcounting for an entry, or you don't, but you don't have it twice once with and once without...

**Comment by poidl** [comment]
> you have the same boot menu entry with and without boot

Not sure what you refer to as "same", one is called `arch-v1`, one is called `arch-v2`. So it's two versions, each of which points to a different kernel/initramfs/rootfs. But yes, the "arch-" part is the same.

I read the man page about boot counting:

from [https://www.freedesktop.org/software/systemd/man/systemd-boot.html#Boot%20Counting](url)

> If the menu entry to boot is automatically determined, this means that 'good' or 'indeterminate' entries are generally preferred (as the top item of the menu is the one booted by default), and 'bad' entries will only be considered if there are no 'good' or 'indeterminate' entries left.

I start with an entry file `arch-v1.conf`. After adding a new kernel/initramfs/rootfs (for me, that triplet constitutes a new "version" of my Arch linux deployment), add an entry file `arch-v2+1.conf` .

**Expected behaviour:**

 Expected boot menu is
```
arch-v2+1
arch-v1
```
and it should try booting the first entry. If it fails to boot, then upon reboot the expected menu entry is

```
arch-v1
arch-v2+0-1
```

and it boots into `arch-v1` without user intervention.

**Observed behaviour:**

Observed boot menu is
```
arch-v1
arch-v2+1
```
and it boots from `arch-v2+1`. So far that's fine. If it fails to boot, then upon reboot the menu entry is
```
arch-v1
arch-v2+0-1
```
and now it **again** boots from `arch-v2+0-1`

**Comment by poidl** [comment]
I should say this now has less priority for me, I found a workaround for my using  `bootctl set-oneshot` and `set-default`. See also this link:

https://blog.printk.io/2020/02/systemd-boot-counting-and-boot-complete-target/

**Comment by benjarobin** [comment]
Is it fixed by #15577 ?

**Comment by poettering** [comment]
Seems, so. let's close. Please report back if it doesn't fix it.
